Fix:7939 Limiting to one card per note in CardBrowser

kartikeysaran opened the pull request:
## Purpose / Description
Previously the Cards were repeated because of reverse card and cloze in Cardbrowser as shown below

<img src="https://user-images.githubusercontent.com/55613721/111958171-eefa8600-8b12-11eb-8804-6f771552cb16.jpeg" width="200" height = "400"/>

Now the Cards are limited to one card per note as shown below

<img src="https://user-images.githubusercontent.com/55613721/111958690-8b248d00-8b13-11eb-8b30-3f9d275a5c7d.jpeg" width = "200" height = "400"/>
<img src = "https://user-images.githubusercontent.com/55613721/111958699-8c55ba00-8b13-11eb-8b3f-cd213545c651.jpeg" width = "200" height = "400" />

## Fixes
Fixes #7939 

## Approach
By using findNotes() method we get a list of note id and then by comparing the note id with the note id inside card we limit the card to one per note

## How Has This Been Tested?

This has been tested locally on emulator

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

kartikeysaran commented:
@Arthur-Milchior Do i need to add a seperate option (eg. a Switch button or a Filter Options ) in Cardbrowser for switching from "All Cards" to "Limited Cards" as in this pr whenever you open card browser you will be shown limited cards

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Finder.java:
```diff
@@ -130,7 +130,21 @@ public Finder(Collection col) {
-        return res;
+        List<Long> nid = findNotes(query);
+        List<Long> cid = new ArrayList<>();
+        for ( int i = 0 ; i < nid.size() ; i++ ) {
```
FYI you can do 

```java
for(Long nid: nids) {
...
}
```
and then instead of doing `nids.get(i)` you can directly use the variable `nid`. It makes code shorter and easier to read.
(I assume here that the name of the list is `nids`)

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Finder.java:
```diff
@@ -130,7 +130,21 @@ public Finder(Collection col) {
+            for ( int j = 0 ; j < res.size() ; j++ ) {
+                int count  = 0;
+                if ( nid.get(i).equals( getNoteId(res.get(j))) ){
+                    count++;
```
That is extremely inefficient. Do you know the concept of "big-O" notation? 
Generally, you should avoid having a loop inside another unless you really need it. If you have let's say 50.000 cards for 10.000 notes (realistic for big users), you'll need to do 500.000.000 equality tests and 500.000.000 database query. That will be very slow.

You could do something similar by creating a set of note ids. Each time you see a card, if the note id is in the set you can ignore it, otherwise you add it to the set and add the card to the list. This will ensure that you only have to loop over cards, not over notes. You should look at the HashSet class if you don't know it.

An even better solution would be to adapt `find_cards` so that it returns pairs of `card id` and `note id` so that you don't have to do multiple database query.

Arthur-Milchior reviewed (commented):
Hi,

Thanks a lot for this commit. I had to add a lot of comment, I hope you will find them useful.  I admit that what you did work, but the code itself still require corrections. There are some problems with styles, like using `cid` instead of `cids` for a list, which is not our standards. Indeed "s" indicates plural. In java you should usually loop over list elements and not indices, unless required (May I ask how much experience you have with java?)

And finally, for a big collection, the code will be very slow. I've got 200 000 cards, and looping over all of them would take quite a lot of time.


Finally, to answer your question, I would indeed like a switch, because some users may still prefer to see every cards; it also has some advantages.

kartikeysaran commented:
Added Menu Action Item (check button) for displaying limited cards
Updated the findCards function taking time complexity into account
Updated the _query function to return note_Ids as well as card_Ids
Updated UI looks like 
<img src="https://user-images.githubusercontent.com/55613721/112296752-1db36080-8cbb-11eb-90c2-e46707e5c75f.gif" width="200" height="400">
@Arthur-Milchior

Arthur-Milchior commented on AnkiDroid/src/main/res/menu/card_browser.xml:
```diff
@@ -14,6 +14,11 @@
+    <item
+        android:id="@+id/action_limited_search"
+        android:checkable="true"
+        android:title="Show Limited Cards"/>
```
Please use a string ressource, as done in the items above. Otherwise we can't translate the string.

I'd state "Show one card per note" but that's up to discussion with other maintainers

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Finder.java:
```diff
@@ -112,42 +113,49 @@ public Finder(Collection col) {
         if (rev) {
             Collections.reverse(res);
         }
-        List<Long> nid = findNotes(query);
```
Thank you for removing this code. However, I'd also like you to remove it from original commit. This way we don't introduce code in history to remove it.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Finder.java:
```diff
@@ -112,42 +113,49 @@ public Finder(Collection col) {
+                    if (isCancelled(task)) {
+                        return new ArrayList<>(0);
+                    }
+                    if ( !nids.contains(cur.getLong(1)) ) {
```
There is a lot of code duplication. Do you think you could remove it by putting the `if` here instead of outside the try?

kartikeysaran replied:
if we put if here instead of outside try and catch then each time when we get a card id we have to check the ``` boolean singleCardByNote ```  which will increase the time complexity as well , rather we can place the if before the ``` while( cur.moveToNext ) ``` this will restrict code duplication to an extent

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Finder.java:
```diff
@@ -76,33 +76,34 @@ public Finder(Collection col) {
 
     @CheckResult
-    private List<Long> _findCards(String query, Object _order, CollectionTask.PartialSearch task) {
+    private List<Long> _findCards(String query, Object _order, CollectionTask.PartialSearch task, boolean limitedCards) {
```
I'd like you to use a more descriptive variable name. `limitedCards` can mean a lot of things. Javadoc will probably be useful. But even with javadoc, I believe that something such as `singleCardByNote` would be more clear

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:
```diff
@@ -1476,6 +1485,17 @@ private void searchCards() {
+            } else {
+                limitedCards = false;
+            }
+        } catch (Throwable e) {
```
Why do you need to catch this? 
I don't expected that isChecked can fail, can it? In any case, you can catch something more specific than just "throwable"

kartikeysaran replied:
Since it the ``` mActionMenu.finditem() ``` can return null if not initialized , i have catched the exception in the new commit

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:
```diff
@@ -1476,6 +1485,17 @@ private void searchCards() {
+        boolean limitedCards;
+        try{
+            if (mActionBarMenu.findItem(R.id.action_limited_search).isChecked()) {
+                limitedCards = true;
```
Instead of a `if`
you could do `limiteCards = mActionBarMenu.findItem(R.id.action_limited_search).isChecked()`

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:
```diff
@@ -1180,7 +1180,16 @@ public boolean onOptionsItemSelected(MenuItem item) {
+        } else if (itemId == R.id.action_limited_search) {
+            if( mActionBarMenu.findItem(R.id.action_limited_search).isChecked() ){
+                mActionBarMenu.findItem(R.id.action_limited_search).setChecked(false);
+                searchCards();
```
Please move `searchCards` outside of the if/else, to avoid code duplication

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:
```diff
@@ -1180,7 +1180,16 @@ public boolean onOptionsItemSelected(MenuItem item) {
             return true;
+        } else if (itemId == R.id.action_limited_search) {
+            if( mActionBarMenu.findItem(R.id.action_limited_search).isChecked() ){
+                mActionBarMenu.findItem(R.id.action_limited_search).setChecked(false);
```
```java
bool wasChecked = mActionBarMenu.findItem(R.id.action_limited_search).isChecked();
mActionBarMenu.findItem(R.id.action_limited_search).setChecked(wasChecked);
``` 
would work as well and be shorter

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Finder.java:
```diff
@@ -388,9 +396,9 @@ private static String _query(String preds, String order) {
         String sql;
         if (!preds.contains("n.") && !order.contains("n.")) {
-            sql = "select c.id from cards c where ";
+            sql = "select c.id , c.nid from cards c where ";
```
I'm sorry, but getting data from the database is very often a bottleneck. So getting nid when we don't need it seems a bad idea.

Actually if you want to select a single card by note you could just do
```sql
select MIN(c.id) from cards c where 
....
group by c.nid```
This ensure that for each note you'd get a single card, the first one

kartikeysaran replied:
This would return a single card id

Arthur-Milchior replied:
Unless i got it wrong, it would get a single card id by note id, that what `group by ` does. You can use sqlitebrowser for example to check.

kartikeysaran replied:
``` select MIN(c.id) from cards c where ``` would return the card with min card id which can only be one and then further we are grouping it by ``` nid ``` , What i mean to say is first the select part would execute then the grouping would be performed the select part would return a single card id to us

Arthur-Milchior replied:
Can you please let me know whether you actually opened a database tool, such as sqlitebrowser, the debugger in anki, or whatever you like and tested it ? 

I just double checked right now with sqlitebrowser on a test collection.

`select min(id) from cards group by nid`
shows 1587 card id. I have 1587 notes in my collection, and >4000 cards. So I really really believe that you're wrong here. Furthermore, according to https://sqlite.org/lang_select.html#resultset

> If the SELECT statement is an aggregate query with a GROUP BY clause, then each of the expressions specified as part of the GROUP BY clause is evaluated for each row of the dataset. Each row is then assigned to a "group" based on the results; rows for which the results of evaluating the GROUP BY expressions are the same get assigned to the same group

kartikeysaran replied:
> rthermore, a

i have tested it in anki debugger ,will recheck it

kartikeysaran replied:
@Arthur-Milchior Yes , you were right , Sorry for the mistakes . By mistake i passed ``` group by ```  after the order which resulted in 0 results I have made commits with the correction and the updates now we are not fetching the data that we dont need i.e. ``` note id ``` , Thankyou for help ! :)

Arthur-Milchior replied:
No problem. 

I am sorry, but since I don't know you, I've really no idea what is your level, whether you knew group by, big-O notation, why some code is faster than some other. So I also don't know at which level my explanation should be. Not sure whether it's clear why this last version seems better to me than the initial one.

At least, I guess it was my mistake, because I believed it would be a good first issue and didn't really thought it would require multiple rewritting

kartikeysaran replied:
It is clear to me why this is better than that because in more complex cases where we have a large number of cards it would be relatively quicker , during my initial commit i didn't had much knowledge of code base but with time i am becoming used to it and much thanks to you in this .

Arthur-Milchior reviewed (commented):
Thanks a lot for corrections.

I'd love, please, for you to change your commit, so that we avoid introducing code we then remove. So squashing would be good.
I woud also love to have a commit for background code and one to change the interface. This allow to read and review both changes independtly and easily

david-allison-1 reviewed (commented):
Wrote this a while back and forgot to submit, if these aren't relevant any more, please mark as resolved

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:
```diff
@@ -1476,6 +1481,13 @@ private void searchCards() {
+        boolean singleCardByNote;
+        try{
+            singleCardByNote = mActionBarMenu.findItem(R.id.action_limited_search).isChecked();
+        } catch ( Exception e) {
```
Please do not catch `Exception`. Normally, you should not have anything to catch here, and if there is really something you need to catch, then catch just the required exception class

kartikeysaran replied:
Will update the code with this

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/libanki/Finder.java:
```diff
@@ -370,28 +370,46 @@ public void add(String txt, boolean wrap) {
-        // combine with preds
-        if (!TextUtils.isEmpty(preds)) {
-            sql += "(" + preds + ")";
+        if (singleCardByNote) {
```
There is a lot of code duplication. I really would prefer to avoid it please. Instead, you can use a variable `String select` whose value is either `c.id` or `min(c.id)` and another variable `String groupBy` whose value would be either `""` or `"group by c.nid"`

Arthur-Milchior commented:
do you know git ? Because at some point I'd really like you to rebase and make atomic commits

kartikeysaran commented:
> do you know git ? Because at some point I'd really like you to rebase and make atomic commits

Yes i do !

david-allison-1 commented:
Lint is failing, tests are failing

```
  /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:1504: Error: Explicitly logging exception message is redundant [TimberExceptionLogging]
              Timber.e(e.getMessage());
              ~~~~~~~~~~~~~~~~~~~~~~~~
```

https://github.com/ankidroid/Anki-Android/pull/8314/checks?check_run_id=2209292782

kartikeysaran commented:
> Lint is failing, tests are failing
> 
> ```
>   /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:1504: Error: Explicitly logging exception message is redundant [TimberExceptionLogging]
>               Timber.e(e.getMessage());
>               ~~~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> https://github.com/ankidroid/Anki-Android/pull/8314/checks?check_run_id=2209292782

Will update the code

Arthur-Milchior commented:
I’m really sorry, there are changes in upstream I didn’t know about at all :(
The ui will have a toggle for note or cards. In note modes, no card column will be shown (e.g due date, question, answer). I don’t know what is done with suspending, whether it applies to note or is disabled.

So using find_note is fine, GROUP BY is useless, and the columns would be more complex to deal with

kartikeysaran commented:
> I’m really sorry, there are changes in upstream I didn’t know about at all :(
> The ui will have a toggle for note or cards. In note modes, no card column will be shown (e.g due date, question, answer). I don’t know what is done with suspending, whether it applies to note or is disabled.
> 
> So using find_note is fine, GROUP BY is useless, and the columns would be more complex to deal with

So should i use ``` findNotes ``` when toggling to note mode and show two columns the front card and the back card ?

Arthur-Milchior commented:
> So should i use `findNotes` when toggling to note mode and show two columns the front card and the back card ?

"Front card" and "back cards" are not terms that exists in anki. A note could have a thousand cards theoretically (Even if it's rare normally)

At this point, I believe the best thing to do is to stop for some time and wait to see what Anki will do. It's going to be a upstream feature and then we went to copy it, but it seems it's not yet totally fixed there

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/Finder.java:
```diff
@@ -76,40 +76,41 @@ public Finder(Collection col) {
     @CheckResult
-    public List<Long> findCards(String query, String _order) {
-        return _findCards(query, _order);
+    public List<Long> findCards(String query, String _order, boolean singleCardByNote) {
```
Is it feasible to perform this filtering in the UI, rather than in the finder? This isn't going to be compatible with the Rust migration (DB v16) in AnkiDroid 2.16

kartikeysaran replied:
Yes it is possible !

Arthur-Milchior replied:
As I tried to explain above, I don't think that this is relevant at all anymore. Mostly because this feature is not yet stable in upstream, and until they finally figure out the exact way they'll deal with it and we see users feedback, working on porting it is probably work that will need to be changed soon
